### ISSUE TITLE
"I Can't Sleep in Silence" crashes when run with 3.5.0.12> build

### ISSUE THREAD

**Issue by necros2k7** [adventuregamestudio/ags#1840]
**Describe the bug**
https://www.adventuregamestudio.co.uk/site/games/game/2031/ game crashes after pressing Start in main menu

**AGS Version**
3.5.0.24. 3.5.0.30, 3.5.1.23, 3.6.0.38
**Game**
https://www.adventuregamestudio.co.uk/site/games/game/2031/ 

**To Reproduce**
1. start game 
2. press Start in main menu

**Expected behavior**
Game start 

**Screenshots**
Illegal exception

An exception 0xC0000005 occurred in ACWIN.EXE at EIP = 0x004C1E93; program pointer is +1100, ACI version 3.5.0.24, gtags (33,2)

**Desktop (please complete the following information):**
 - OS: Win
 - Version 11
 
latest build 3.6.x gives Out of memory error

**Comment by ivan-mogilko** [comment]
This is related to the engine not handling snowrain plugin correctly, because game drawing is done somewhat differently since 3.5.0.

**Comment by ivan-mogilko** [comment]
Our snowrain plugin rewrite from this repository does not crash the game, but it draws the rain only on the top half of the screen.

**Comment by necros2k7** [comment]
@ivan-mogilko , true only half screen

**Comment by ivan-mogilko** [comment]
I'm currently planning to make some changes, if not making this work with the older plugin, but at least to make it not crash.
But this is bit lower priority than another serious issue in 3.6.0 (#1838), otoh is somehow connected, being related to the same part of the graphics code. So maybe I will return to this after #1838.

**Comment by ivan-mogilko** [comment]
I will note my findings so far.
The main reason of this crash is following:
* plugin sets its own memory buffer for rendering, by calling IAGSEngine::SetVirtualScreen, which in turn calls [SDLRendererGraphicsDriver::SetMemoryBackBuffer](https://github.com/adventuregamestudio/ags/blob/5d051d614cafe2a4f73cc4399eeb78cb38d8ed2d/Engine/gfx/ali3dsw.cpp#L587).
* In SetMemoryBackBuffer the renderer deletes all existing "sub-surfaces" previously prepared from the existing virtual screen.
 
The problem here is that the callback, from which plugin calls SetVirtualScreen, is made in the middle of rendering a batch, and one of the deleted surfaces is already stored in a temporary local variable, which is being used while rendering current batch. So, when the callback is over, and renderer continues to draw regular sprites, it tries to do so on a deleted bitmap.

There's definitely a logical mistake here, in how these "batch surfaces" are managed.

But then there's something else too. Suppose the surfaces are not deleted (the part of the code may be commented out, as a test), the engine crashes anyway some time later, when doing ["Fade in" effect](https://github.com/adventuregamestudio/ags/blob/5d051d614cafe2a4f73cc4399eeb78cb38d8ed2d/Engine/gfx/ali3dsw.cpp#L771).

At this moment I was not able to understand why exactly this second crash is happening. But I found that it may be prevented by not clearing sprite batches in [SDLRendererGraphicsDriver::ResetAllBatches](https://github.com/adventuregamestudio/ags/blob/5d051d614cafe2a4f73cc4399eeb78cb38d8ed2d/Engine/gfx/ali3dsw.cpp#L377). Maybe there's another problem with something getting deleted in a wrong time.

**Comment by ivan-mogilko** [comment]
It took me a while to realize the real mistake I did in 3.5.0, but finally seem to make a proper fix. This is in release-3.5.1 branch for now, will port to 3.6.0 soon:
5a6fe4a216d0c90c1c1a3362d562f7b41f4c3e74
aba50ee84f28c66d1deaf71fa1c31e5b478b13b8
(more info is in commit comments)

**Comment by ivan-mogilko** [comment]
Should be fixed in 3.6.0 now, related commits:
5dd078961020da6478347dac01dea729471333d1
09143ea7fbf8474f78932116ae1c81ceff4de95a
